### PR TITLE
[glib] Use meson build system.

### LINKED ISSUE #11476: [glib] Upgrade glib and use meson build system.
glib  port is 3 years old and use a custom CMakeSystem.
glib now uses meson as a build system. Since vcpkg now supports this build system it would be better to use it to ease maintainability.

### PR DESCRIPTION
**Pull request by vlj**
**Describe the pull request**

- What does your PR fix? Fixes https://github.com/microsoft/vcpkg/issues/11476

- Which triplets are supported/not supported? Have you updated the CI baseline?
Normally all previous triplets are supported (at least tested x64 and ported arm patches, hopefully mac works out of the box)

About the PR:
It ditches CMakeLists.txt (apparently some files like gfilemonitor were not even included by glib) and use meson instead. To preserve backward compatibility I added some targets.cmake files which are copied after installation, they allow others ports to maintain compatibility.

There is no more config.h.win32.in with meson but several .in files, I think that it's now responsible to handle arm64.

### CODE REVIEW

**Comment by vlj** [comment]
I'm using this PR to see if things build with all platforms.

**Comment by vlj** [comment]
How can I fix linux build ? It complains that meson is not installed by apt.

**Comment by vlj** [comment]
arm64 fails because cross compilation is not enabled in meson so far... And it's the same for fribidi.

**Comment by c72578** [comment]
> 
> How can I fix linux build ? It complains that meson is not installed by apt.

Meson needs to be installed on the Linux build host.
```console
CMake Error at scripts/cmake/vcpkg_find_acquire_program.cmake:367 (message):
  Could not find meson.  Please install it via your package manager:

      sudo apt-get install meson
```

**Comment by JackBoosY** [comment]
I think we should fix meson cross compilation first.

**Comment by Voskrese** [comment]
> 
> sudo apt-get install meson
> 

not latest version

**Comment by vlj** [comment]
> 
> Meson needs to be installed on the Linux build host.
> 

But how can I do that ? Is there a script that control what is installed on azure agents ?

**Comment by vlj** [comment]
I changed include pathes, meson build added a glib-2.0/ prefix to very glib file but every packages in vcpkg expected glib includes to be in include/

**Comment by vlj** [comment]
By the way the osx64 build fails but it doesn't tell why ?

**Comment by Voskrese** [comment]
https://cmake.org/cmake/help/latest/variable/CMAKE_HOST_UNIX.html

https://github.com/microsoft/vcpkg/blob/c4937039b0704c711dff11ffa729f1c105b20e42/scripts/cmake/vcpkg_find_acquire_program.cmake#L225

```
    if(CMAKE_HOST_WIN32 AND CMAKE_HOST_UNIX)
      set(SCRIPTNAME meson.py)
    else()
      set(SCRIPTNAME meson)
    endif()
```

...

```
    if(CMAKE_HOST_UNIX)
      find_package(Python3 COMPONENTS Interpreter)
      set(PYTHON3 ${PYTHON3_EXECUTABLE})
    else()
```

?

**Comment by Neumann-A** [comment]
> vlj force-pushed the vlj:glib-meson branch from a2d7b9f to 5254d9b 3 hours ago
ah man ..... and i was just doing that
```
Building package glib[core]:arm64-windows... done
Installing package glib[core]:arm64-windows...
Installing package glib[core]:arm64-windows... done
Elapsed time for package glib:arm64-windows: 2.114 min
```

**Comment by Neumann-A** [comment]
https://github.com/Neumann-A/vcpkg/tree/fix_meson_cross

requires
https://github.com/mesonbuild/meson/pull/7190/files

File: ports/glib/CONTROL [hunk 1]
```diff
@@ -1,8 +1,8 @@
 Source: glib
-Version: 2.52.3-14-6
+Version: 2.64.3
 Homepage: https://developer.gnome.org/glib/
 Description: Portable, general-purpose utility library.
-Build-Depends: zlib, pcre, libffi, gettext, libiconv
+Build-Depends: zlib, pcre, libffi, gettext, libiconv, tool-meson
```
Comment (c72578) [inline ports/glib/CONTROL:5]: Could you add `dirent.h` to `Build-Depends`, which removes the following error message in **`meson-log.txt`**:
```console
testfile.c(4): fatal error C1189: #error:  "Header 'dirent.h' could not be found"
Has header "dirent.h" : NO
```

**Comment by Neumann-A** [comment]
```
Run-time dependency libpcre found: NO (tried pkgconfig and cmake)
Library pcred found: NO
```

```
Run-time dependency libelf found: NO (tried pkgconfig and cmake)
Library elf found: NO
```

**Comment by c72578** [comment]
How about adding the following to portfile.cmake:
```cmake
if(CMAKE_HOST_WIN32)
set(MESON_COMMON_CFLAGS "${MESON_COMMON_CFLAGS} /I${CURRENT_INSTALLED_DIR}/include")
set(MESON_DEBUG_LDFLAGS "${MESON_DEBUG_LDFLAGS} /LIBPATH:${CURRENT_INSTALLED_DIR}/debug/lib")
set(MESON_RELEASE_LDFLAGS "${MESON_RELEASE_LDFLAGS} /LIBPATH:${CURRENT_INSTALLED_DIR}/lib")
endif()
```
This will enable e.g. the following (see meson-log.txt):
```console
Has header "dirent.h" : YES 
Library pcre found: YES
```

**Comment by vlj** [comment]
Are these defines not done by vcpkg_configure_meson ?

**Comment by c72578** [comment]
> 
> 
> Are these defines not done by vcpkg_configure_meson ?

Not yet ...

**Comment by Neumann-A** [comment]
> Are these defines not done by vcpkg_configure_meson ?

Non of the configure scripts do that except `vcpkg_configure_make` which uses the environment for that. The correct thing to do here is to use/create a `libpcre.pc` file for pcre

**Comment by vlj** [comment]
I added these defines to portfile.cmake, and fixed the port (tools are copied to ports, includes are now in include/ and not in include/glib-2.0 for compatibility with existing ports...).
I discovered that atk won't build anymore because it expects a tool, glib-genmarshal, to be in exe format while it has been moved to python with latest glib. I'm working on it.

**Comment by vlj** [comment]
It seems atk is fixed, and I added dirent to build depends.

File: ports/glib/portfile.cmake [hunk 2]
```diff
@@ -36,22 +36,62 @@ vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
...
+    file(GLOB DEBUG_EXE ${CURRENT_PACKAGES_DIR}/debug/bin/*.exe)
+    if(DEBUG_EXE)
+        file(REMOVE ${DEBUG_EXE})
+    endif()
+    foreach(TOOL ${TOOLS})   
+        file(RENAME ${CURRENT_PACKAGES_DIR}/bin/${TOOL} ${CURRENT_PACKAGES_DIR}/tools/${PORT}/${TOOL})
+    endforeach(TOOL)
+endif()
+
+vcpkg_copy_tool_dependencies(${CURRENT_PACKAGES_DIR}/tools/${PORT})
```
Comment (Neumann-A) [inline ports/glib/portfile.cmake:89]: use `vcpkg_copy_tools` with `autoclean`

File: ports/glib/fix-include-path.patch [hunk 3]
```diff
@@ -0,0 +1,117 @@
+Index: glib/gio/meson.build
+===================================================================
+--- glib.orig/gio/meson.build
++++ glib/gio/meson.build
+@@ -173,7 +173,7 @@ gnetworking_h_conf.set('NAMESER_COMPAT_I
+ 
+ gnetworking_h = configure_file(input : 'gnetworking.h.in',
+                                output : 'gnetworking.h',
+-                               install_dir : join_paths(get_option('includedir'), 'glib-2.0/gio'),
++                               install_dir : join_paths(get_option('includedir'), 'gio'),
```
Comment (Neumann-A) [inline ports/glib/fix-include-path.patch:10]: which ports fail with glib-2.0? Did you also change the interface_include_dir in the unofficial targets?
I don't feel like this is the right solution

Comment (vlj) [reply ports/glib/fix-include-path.patch:10]: There was a dozen ports failing like glibmm, libcroco, libxdgmime...
I investigated glibmm, it expected glib.h to be located in include/glib.h and not in include/glib-2.0/glib.h
Others ports seems fixed by moving every includes too.
glibmm doesn't use find_package, it uses find_path. I don't know what is the best solution, by moving file in the location previous ports expect them, or by fixing every ports.

**Comment by vlj** [comment]
I use vcpkg_copy_tools now.

Note that some tools are python file without any extension (not .exe, not even .py) so I still manually renamed them.

File: ports/glib/portfile.cmake (hunk 2 above)
Comment (vlj) [reply ports/glib/portfile.cmake]: done (for exe tools)

File: ports/glib/fix-include-path.patch (hunk 3 above)
Comment (vlj) [reply ports/glib/fix-include-path.patch:10]: I would really like to be as atomic as possible atm : first get glib to properly build with meson on all supported platforms, then switch some big lib relying on meson (cairo, pango, atk, ...) to vcpkg_meson, and then use the original include tree layout.

Comment (c72578) [reply ports/glib/fix-include-path.patch:10]: > 
> 
> There was a dozen ports failing like glibmm, libcroco, libxdgmime...
> I investigated glibmm, it expected glib.h to be located in include/glib.h and not in include/glib-2.0/glib.h
> Others ports seems fixed by moving every includes too.
> glibmm doesn't use find_package, it uses find_path. I don't know what is the best solution, by moving file in the location previous ports expect them, or by fixing every ports.


The solution should definitely be, to have the include files there, where they are intended and designed to be, under:
`include/glib-2.0/`
This is where upstream glib puts them, and where the include files are typically found and expected to be (under Linux OS, mingw32, mingw64, ...)

Comment (vlj) [reply ports/glib/fix-include-path.patch]: I'm trying to fix the ports then.

**Comment by c72578** [comment]
> 
> Non of the configure scripts do that except `vcpkg_configure_make` which uses the environment for that. The correct thing to do here is to use/create a `libpcre.pc` file for pcre

A PR, which adds `pkgconfig` files for `pcre` has been created: https://github.com/microsoft/vcpkg/pull/11564

**Comment by vlj** [comment]
I reverted the include path to their original location.
There is a header in lib/glib-2.0/include/glibconfig.h, I leave it there too but I don't know if it's the right way to do. I think some others package may expect it there if they are moved to meson.

**Comment by c72578** [comment]
> 
> 
> I reverted the include path to their original location.
> There is a header in lib/glib-2.0/include/glibconfig.h, I leave it there too but I don't know if it's the right way to do. I think some others package may expect it there if they are moved to meson.

I found some explanation concerning `lib/glib-2.0/include/glibconfig.h` here:
https://github.com/libvips/libvips/issues/909#issuecomment-374866784

**Comment by vlj** [comment]
I fixed most build error except for atkmm for now (it's a msbuild like ports so I need some time to figure out how to fix it)

**Comment by Voskrese** [comment]
if do, then the whole GNOME, under MESON

**Comment by vlj** [comment]
?

**Comment by Neumann-A** [comment]
@vlj He is saying to also switch the other ports to meson. But that is a different PR in my opinion

**Comment by c72578** [comment]
> 
> 
> I fixed most build error except for atkmm for now (it's a msbuild like ports so I need some time to figure out how to fix it)

@vlj Here is a workaround for **`atkmm`**. Please have a look.
However, I guess in the future we will also switch to meson for atkmm and update it to a newer version ...

```patch
diff --git a/ports/atkmm/CONTROL b/ports/atkmm/CONTROL
index caf3abb43..3ffe29ba2 100644
--- a/ports/atkmm/CONTROL
+++ b/ports/atkmm/CONTROL
@@ -1,5 +1,5 @@
 Source: atkmm
-Version: 2.24.2-2
+Version: 2.24.2-3
 Homepage: https://www.gtkmm.org
 Description: atkmm is the official C++ interface for the ATK accessibility toolkit library. It may be used, for instance, by user interfaces implemented with gtkmm.
 Build-Depends: glib, gettext, atk, glibmm
diff --git a/ports/atkmm/fix_properties.patch b/ports/atkmm/fix_properties.patch
index a03924386..41e2ea404 100644
--- a/ports/atkmm/fix_properties.patch
+++ b/ports/atkmm/fix_properties.patch
@@ -16,7 +16,7 @@ index 421f5c3..0275e6e 100644
    <ItemDefinitionGroup>
      <ClCompile>
 -      <AdditionalIncludeDirectories>.\atkmm;..\atk;$(GlibEtcInstallRoot)\include\glibmm-2.4;$(GlibEtcInstallRoot)\lib\glibmm-2.4\include;$(GlibEtcInstallRoot)\include\sigc++-2.0;$(GlibEtcInstallRoot)\lib\sigc++-2.0\include;$(GlibEtcInstallRoot)\include\atk-1.0;$(GlibEtcInstallRoot)\include\glib-2.0;$(GlibEtcInstallRoot)\lib\glib-2.0\include;$(GlibEtcInstallRoot)\include;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
-+      <AdditionalIncludeDirectories>.\atkmm;..\atk;.;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
++      <AdditionalIncludeDirectories>.\atkmm;..\atk;.;$(GlibIncludeDir1);$(GlibIncludeDir2);%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
        <ForcedIncludeFiles>msvc_recommended_pragmas.h;%(ForcedIncludeFiles)</ForcedIncludeFiles>
        <MultiProcessorCompilation>true</MultiProcessorCompilation>
        <AdditionalOptions>/d2Zi+ %(AdditionalOptions)</AdditionalOptions>
diff --git a/ports/atkmm/portfile.cmake b/ports/atkmm/portfile.cmake
index f8e9df294..6b58c1704 100644
--- a/ports/atkmm/portfile.cmake
+++ b/ports/atkmm/portfile.cmake
@@ -1,5 +1,3 @@
-include(vcpkg_common_functions)
-
 vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)
 
 vcpkg_download_distfile(ARCHIVE
@@ -26,6 +24,12 @@ vcpkg_build_msbuild(
     PROJECT_PATH ${SOURCE_PATH}/MSVC_Net2013/atkmm.sln
     TARGET atkmm
     PLATFORM ${VS_PLATFORM}
+    OPTIONS
+    # Remark: This is a workaround for now. When updating atkmm to a newer version, meson should be used instead
+    # Pass the two required glib include directories separately, as semicolons are challenging under cmake
+    # GlibIncludeDir1 and GlibIncludeDir2 are used by fix_properties.patch
+        /p:GlibIncludeDir1="${CURRENT_INSTALLED_DIR}/include/glib-2.0"
+        /p:GlibIncludeDir2="${CURRENT_INSTALLED_DIR}/lib/glib-2.0/include"
     USE_VCPKG_INTEGRATION
 )
 
```

**Comment by vlj** [comment]
Sorry for the delay, I added you patch thanks !

**Comment by vlj** [comment]
I needed to fix glib gdbus-codegen because it hardcoded path to package.

**Comment by vlj** [comment]
I fixed gtkmm.

**Comment by vlj** [comment]
It seems like all dependent project are fixed (at least on windows).
Can't wait to get the meson PRs and this one merged so that all project can be switched to meson.

**Comment by Neumann-A** [comment]
mind trying out #12945?

**Comment by vlj** [comment]
What do you mean ?

**Comment by JackBoosY** [comment]
Could you please solve these regressions?

**Pull request closed without merging**